target/scripts/helper-functions.sh

```diff
@@ -12,6 +12,37 @@ function errex
   exit 1
 }
 
+PANIC_NO_ENV='no-env'
+PANIC_NO_FILE='no-file'
+PANIC_INVALID_VALUE='invalid-value'
+function dms_panic
```

Comment:
I'd like to see some documentation for this function :)

Reply:
Sure! Did you want inline documentation in this file, or did you mean our contributor docs?

Reply:
Inline or above the function:)

Reply:
We already have and use `_shutdown()`.

Shouldn't we enhance this function, instead of introducing a new one, that is basically doing the same?

E.g. afaik `_shutdown()` is never called without a prior `_notify()` call. So why not move the `_notify` call inside `_shutdown()` and make the notify message a parameter for `_shutdown()`? That's basically what _exit_with_message() in this PR does.

This leads to, why do we need `dms_panic()` when there is already `_shutdown()` (which could be extended if needed).

Reply:
> This leads to, why do we need `dms_panic()`

It just provides convenience messages for certain types of errors to avoid repetition (which'd likely diverge a bit).

> when there is already `_shutdown()`

Shutdown is doing a `kill 1` instead of `exit 1`, what is the difference? Panic seems more similar to `errex` (error exit?), as it's meant to stop/exit immediately, whereas shutdown usually implies more of a process with graceful shutdown orchestration.

I chose "panic" as akin to kernel panic, or panic in Rust. It additionally uses the `_notify` with "fatal" to indicate something has gone quite wrong/broke, as in show stopping not a recoverable/ignorable error.

Happy to rework it if you like or integrate with `_shutdown` or `errex`.

Reply:
> Inline or above the function:)

Done, above the function, a tad verbose though :sweat_smile:

Reply:
> Shutdown is doing a kill 1 instead of exit 1, what is the difference?

`kill 1` will shutdown the container, by sending a SIGTERM signal to PID 1 (init process).
`exit 1` in the `start-mailserver.sh` context, will just exit the script. Supervisor will [not restart](https://github.com/docker-mailserver/docker-mailserver/blob/2bf24e4c08dd652f9f4d2bdef639978ad7406d30/target/supervisor/conf.d/supervisor-app.conf#L11) that service. This leads to a container, where only the init process + supervisor (without any services) is running.

So `exit 1` is probably not what you want.

I would go with `_shutdown()` and enhance it (or "merge" with your current solution _dms_panic), to fit also your needs.